include/slam_toolbox/laser_utils.hpp

```diff
@@ -75,17 +75,24 @@ class LaserMetadata
 class LaserAssistant
 {
 public:
+  template <class NodeT>
   LaserAssistant(
-    rclcpp::Node::SharedPtr node, tf2_ros::Buffer * tf,
-    const std::string & base_frame);
+    NodeT && node, tf2_ros::Buffer *tf,
+    const std::string &base_frame)
+    : LaserAssistant(node->get_node_logging_interface(), node->get_node_parameters_interface(), tf, base_frame) {}
```

Comment:
I fail to see why a second constructor is required

Reply:
There are several reasons for this change and my changes ware influenced by <https://github.com/ros-perception/image_common/pull/167#issuecomment-826995226>.

- template cannot be separeted into source and header
  - <https://isocpp.org/wiki/faq/templates#templates-defn-vs-decl>
- `rclcpp_lifecycle::LifecycleNode` does not have the same base class as rclcpp::Node. So, it could not be cast.
  - Node definition : <https://github.com/ros2/rclcpp/blob/humble/rclcpp/include/rclcpp/node.hpp#L77>
  - LifecycleNode defintion : <https://github.com/ros2/rclcpp/blob/humble/rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp#L123> <https://github.com/ros2/rclcpp/blob/humble/rclcpp_lifecycle/include/rclcpp_lifecycle/node_interfaces/lifecycle_node_interface.hpp#L42>

<details> <summary> undefined error when splitting into source and header </summary> 

```diff
diff --git a/include/slam_toolbox/laser_utils.hpp b/include/slam_toolbox/laser_utils.hpp
index de31019..e61ca77 100644
--- a/include/slam_toolbox/laser_utils.hpp
+++ b/include/slam_toolbox/laser_utils.hpp
@@ -77,13 +77,8 @@ class LaserAssistant
 public:
   template <class NodeT>
   LaserAssistant(
-    NodeT && node, tf2_ros::Buffer *tf,
-    const std::string &base_frame)
-    : LaserAssistant(node->get_node_logging_interface(), node->get_node_parameters_interface(), tf, base_frame) {}
-  LaserAssistant(
-    rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr logging_interface,
-    rclcpp::node_interfaces::NodeParametersInterface::SharedPtr parameters_interface,
-    tf2_ros::Buffer * tf, const std::string & base_frame);
+    std::shared_ptr<NodeT> node, tf2_ros::Buffer *tf,
+    const std::string &base_frame);
   ~LaserAssistant();
   LaserMetadata toLaserMetadata(sensor_msgs::msg::LaserScan scan);
 
diff --git a/src/laser_utils.cpp b/src/laser_utils.cpp
index 3129ce1..f7add4f 100644
--- a/src/laser_utils.cpp
+++ b/src/laser_utils.cpp
@@ -68,12 +68,12 @@ void LaserMetadata::invertScan(sensor_msgs::msg::LaserScan & scan) const
   scan.intensities = temp.intensities;
 }
 
+template <class NodeT>
 LaserAssistant::LaserAssistant(
-  rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr logging_interface,
-  rclcpp::node_interfaces::NodeParametersInterface::SharedPtr parameters_interface,
+  std::shared_ptr<NodeT> node,
   tf2_ros::Buffer * tf,
   const std::string & base_frame)
-: logger_(logging_interface->get_logger()),parameters_interface_(parameters_interface), tf_(tf), base_frame_(base_frame)
+: logger_(node->get_logger()),parameters_interface_(node->get_node_parameters_interface()), tf_(tf), base_frame_(base_frame)
 {
 }
```

```console
--- stderr: slam_toolbox                                        
/usr/bin/ld: libtoolbox_common.so: undefined reference to `laser_utils::LaserAssistant::LaserAssistant<rclcpp_lifecycle::LifecycleNode>(std::shared_ptr<rclcpp_lifecycle::LifecycleNode>, tf2_ros::Buffer*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/merge_maps_kinematic.dir/build.make:312: merge_maps_kinematic] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:765: CMakeFiles/merge_maps_kinematic.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
/usr/bin/ld: libtoolbox_common.so: undefined reference to `laser_utils::LaserAssistant::LaserAssistant<rclcpp_lifecycle::LifecycleNode>(std::shared_ptr<rclcpp_lifecycle::LifecycleNode>, tf2_ros::Buffer*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/async_slam_toolbox_node.dir/build.make:313: async_slam_toolbox_node] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:736: CMakeFiles/async_slam_toolbox_node.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< slam_toolbox [1min 8s, exited with code 2]
```

</details>

---


When dealing with NodeT as it is, I think that the following changes are necessary.

1:  Make the definition header-only.

<details> <summary> example</summary>

```diff
diff --git a/include/slam_toolbox/laser_utils.hpp b/include/slam_toolbox/laser_utils.hpp
index de31019..26c3dcd 100644
--- a/include/slam_toolbox/laser_utils.hpp
+++ b/include/slam_toolbox/laser_utils.hpp
@@ -77,13 +77,13 @@ class LaserAssistant
 public:
   template <class NodeT>
   LaserAssistant(
-    NodeT && node, tf2_ros::Buffer *tf,
+    std::shared_ptr<NodeT> node, tf2_ros::Buffer *tf,
     const std::string &base_frame)
-    : LaserAssistant(node->get_node_logging_interface(), node->get_node_parameters_interface(), tf, base_frame) {}
-  LaserAssistant(
-    rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr logging_interface,
-    rclcpp::node_interfaces::NodeParametersInterface::SharedPtr parameters_interface,
-    tf2_ros::Buffer * tf, const std::string & base_frame);
+    : logger_(node->get_logger()), parameters_interface_(node->get_node_parameters_interface()), tf_(tf), base_frame_(base_frame){
+      // initialize
+      // get_** ...
+      // create_** ...
+    }
   ~LaserAssistant();
   LaserMetadata toLaserMetadata(sensor_msgs::msg::LaserScan scan);
```
</details>

2: Declare functions with explicit types on the source side.

<details> <summary> example </summary>

```diff
diff --git a/include/slam_toolbox/laser_utils.hpp b/include/slam_toolbox/laser_utils.hpp
index de31019..bc6ac76 100644
--- a/include/slam_toolbox/laser_utils.hpp
+++ b/include/slam_toolbox/laser_utils.hpp
@@ -25,6 +25,7 @@
 #include <vector>
 
 #include "rclcpp/rclcpp.hpp"
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
 #include "slam_toolbox/toolbox_types.hpp"
 #include "tf2/utils.h"
 
@@ -77,13 +78,8 @@ class LaserAssistant
 public:
   template <class NodeT>
   LaserAssistant(
-    NodeT && node, tf2_ros::Buffer *tf,
-    const std::string &base_frame)
-    : LaserAssistant(node->get_node_logging_interface(), node->get_node_parameters_interface(), tf, base_frame) {}
-  LaserAssistant(
-    rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr logging_interface,
-    rclcpp::node_interfaces::NodeParametersInterface::SharedPtr parameters_interface,
-    tf2_ros::Buffer * tf, const std::string & base_frame);
+    std::shared_ptr<NodeT> node, tf2_ros::Buffer *tf,
+    const std::string &base_frame);
   ~LaserAssistant();
   LaserMetadata toLaserMetadata(sensor_msgs::msg::LaserScan scan);
 
diff --git a/src/laser_utils.cpp b/src/laser_utils.cpp
index 3129ce1..e133299 100644
--- a/src/laser_utils.cpp
+++ b/src/laser_utils.cpp
@@ -68,13 +68,20 @@ void LaserMetadata::invertScan(sensor_msgs::msg::LaserScan & scan) const
   scan.intensities = temp.intensities;
 }
 
+template LaserAssistant::LaserAssistant(
+    std::shared_ptr<rclcpp::Node> node, tf2_ros::Buffer *tf, const std::string &base_frame);
+template LaserAssistant::LaserAssistant(
+    std::shared_ptr<rclcpp_lifecycle::LifecycleNode> node, tf2_ros::Buffer *tf, const std::string &base_frame);
+template <class NodeT>
 LaserAssistant::LaserAssistant(
-  rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr logging_interface,
-  rclcpp::node_interfaces::NodeParametersInterface::SharedPtr parameters_interface,
-  tf2_ros::Buffer * tf,
-  const std::string & base_frame)
-: logger_(logging_interface->get_logger()),parameters_interface_(parameters_interface), tf_(tf), base_frame_(base_frame)
-{
+    std::shared_ptr<NodeT> node,
+    tf2_ros::Buffer *tf,
+    const std::string &base_frame)
+    : logger_(node->get_logger()), parameters_interface_(node->get_node_parameters_interface()), tf_(tf), base_frame_(base_frame)
+{
+  // initialize
+  // get_** ...
+  // create_** ...
 }
 
 LaserAssistant::~LaserAssistant()
```
</details>

Alternatively, I can write the same process for Node and LifecycleNode, or replace it directry with  Node with LifecycleNode.  
I would like to confirm which method to use in this repository.

Reply:
Use the node template instead, its a cleaner method. It doesn't need to share a base class to be a template because both implement the same functions.

Reply:
Thank you for confirming.
I will change to use node template. Considering the amout of change from the original code, I will rewrite it with the latter pattern `2: Declare functions with explicit types on the source side.`

I think the class variables will be based on interfaces, because `LifecycleNode::SharedPtr` cannot be handled by `Node::SharedPtr`. (imho it can be resolved by templating of the class itself, but it affects many other functions.)

Reply:
It should make this PR also much more easily readable :-) 

There's not really any custom nodes and alot of places in the ecosystem where we do instantiate them and that's fine since there's only 2 node types. If we start adding more, then it becomes more of a concern

Reply:
Sory, I was getting ahead of myself. should have confirmed earlier. 

---

I'm currently working on fix. I only have Mapper.h and ceres_solver.(cpp|hpp) left, but I can't use template because Configure (in Mapper.h) is `virtual`.   Can I duplicate the function in this case? I would like to know the direction of changing in this repository.  

I'm thinking of either changing the virtual function in mapper.h to node_interface (= this is current PR change) or using template in ceres_solver.hpp for common processing.

<details> <summary>use template in ceres_solver.hpp for common processing. </summary> 

```cpp
# Mapper.h
...
  virtual void Configure(rclcpp::Node::SharedPtr node) = 0;
  virtual void Configure(rclcpp_lifecycle::LifecycleNode::SharedPtr node) = 0;
...

# ceres_solver.hpp
private:
  template<class NodeT>
  void ConfigureT(std::shared_ptr<NodeT> node);
public:
  virtual void Configure(rclcpp::Node::SharedPtr node)
  {
    ConfigureT(node);
  }
  virtual void Configure(rclcpp_lifecycle::LifecycleNode::SharedPtr node)
  {
    ConfigureT(node);
  }
```

</details>

Reply:
>  Can I duplicate the function in this case?

How about just using LifecycleNode within Mapper. That's an internal-only facing SDK to SLAM Toolbox that few people ever look at and fewer would ever significantly modify.

Reply:
> How about just using LifecycleNode within Mapper. That's an internal-only facing SDK to SLAM Toolbox that few people ever look at and fewer would ever significantly modify.

I see. I'll use LifecycleNode simply. 👍